src/interaction/keyframerecording.cpp

```diff
@@ -0,0 +1,359 @@
+    j.at("application").get_to(ts.application);
+    j.at("sequence").get_to(ts.sequenceTime);
+    j.at("simulation").get_to(ts.simulation);
+}
+
+void from_json(const nlohmann::json& j, KeyframeNavigator::CameraPose& pose) {
+    j.at("position").at("x").get_to(pose.position.x);
+    j.at("position").at("y").get_to(pose.position.y);
+    j.at("position").at("z").get_to(pose.position.z);
+
+    j.at("rotation").at("x").get_to(pose.rotation.x);
+    j.at("rotation").at("y").get_to(pose.rotation.y);
+    j.at("rotation").at("z").get_to(pose.rotation.z);
+    j.at("rotation").at("w").get_to(pose.rotation.w);
+
+    j.at("focusNode").get_to(pose.focusNode);
+    j.at("scale").get_to(pose.scale);
+    j.at("followFocusNodeRotation").get_to(pose.followFocusNodeRotation);
+}
+
+void from_json(const nlohmann::json& j, KeyframeRecording::Keyframe& keyframe) {
+    j.at("camera").get_to(keyframe.camera);
+    j.at("timestamp").get_to(keyframe.timestamp);
+}
+
+KeyframeRecording::KeyframeRecording()
+    : properties::PropertyOwner({ "KeyframeRecording", "Keyframe Recording" })
+{}
+
+void KeyframeRecording::newSequence() {
+    _keyframes.clear();
+    _filename.clear();
+}
+
+void KeyframeRecording::addKeyframe(double sequenceTime) {
+    Keyframe keyframe = newKeyframe(sequenceTime);
+
+    auto it = std::find_if(
+        _keyframes.begin(),
+        _keyframes.end(),
+        [sequenceTime](const Keyframe& entry) {
+            return sequenceTime < entry.timestamp.sequenceTime;
+        }
+    );
+    _keyframes.insert(it, keyframe);
+}
+
+void KeyframeRecording::removeKeyframe(int index)
+{
+    if (index < 0 || static_cast<size_t>(index) >(_keyframes.size() - 1)) {
+        LERROR(std::format("Index {} out of range", index));
+        return;
+    }
+    _keyframes.erase(_keyframes.begin() + index);
+}
+
+void KeyframeRecording::updateKeyframe(int index) {
+    if (index < 0 || static_cast<size_t>(index) > (_keyframes.size() - 1)) {
+        LERROR(std::format("Index {} out of range", index));
+        return;
+    }
+    Keyframe old = _keyframes[index];
+    _keyframes[index] = newKeyframe(old.timestamp.sequenceTime);
+}
+
+void KeyframeRecording::moveKeyframe(int index, double sequenceTime) {
+    if (index < 0 || static_cast<size_t>(index) >(_keyframes.size() - 1)) {
+        LERROR(std::format("Index {} out of range", index));
+        return;
+    }
+
+    _keyframes[index].timestamp.sequenceTime = sequenceTime;
+    sortKeyframes();
+}
+
+bool KeyframeRecording::saveSequence(std::optional<std::string> filename) {
+    // If we didn't specify any filename we save the one we currently have stored
+    if (filename.has_value()) {
+        _filename = filename.value();
+    }
+
+    if (_filename.empty()) {
+        LERROR("Failed to save file, reason: Invalid empty file name");
+        return false;
+    }
+
+    nlohmann::json sequence = _keyframes;
+    std::filesystem::path path = absPath("${RECORDINGS}/" + _filename + ".json");
```

Comment:
```suggestion
    std::filesystem::path path = absPath(std::format(${RECORDINGS}/{}.json", _filename));
```

Reply:
Can't be done because "${RECORDINGS} is not compile time constant :'(